File: backend/src/routes/projects.ts
import { Router, Request, Response, NextFunction } from 'express';
import * as projectService from '../services/projectService';
import { authMiddleware } from '../middleware/auth';
import { successResponse, errorResponse } from '../utils/response';
import { z } from 'zod';

const router = Router();
router.use(authMiddleware);

const projectSchema = z.object({
  name: z.string().min(1, 'Name is required'),
  description: z.string().optional(),
});

// GET /projects/stats (Bonus Feature)
router.get('/stats', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const userId = (req as any).user.user_id;
    const { rows } = await require('../db').default.query(`
      SELECT 
        (SELECT COUNT(*) FROM projects WHERE owner_id = $1) as total_projects,
        (SELECT COUNT(*) FROM tasks t JOIN projects p ON t.project_id = p.id WHERE p.owner_id = $1 OR t.assignee_id = $1) as total_tasks,
        (SELECT COUNT(*) FROM tasks t JOIN projects p ON t.project_id = p.id WHERE (p.owner_id = $1 OR t.assignee_id = $1) AND t.status = 'done') as completed_tasks
    `, [userId]);
    return successResponse(res, rows[0]);
  } catch (err) {
    next(err);
  }
});

// GET /projects
router.get('/', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const userId = (req as any).user.user_id;
    const projects = await projectService.getProjectsForUser(userId);
    return successResponse(res, { projects });
  } catch (err) {
    next(err);
  }
});

// POST /projects
router.post('/', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const userId = (req as any).user.user_id;
    const validated = projectSchema.parse(req.body);
    const project = await projectService.createProject(validated.name, validated.description || '', userId);
    return successResponse(res, project, 201);
  } catch (err) {
    next(err);
  }
});

// GET /projects/:id
router.get('/:id', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const projectId = req.params.id;
    const userId = (req as any).user.user_id;
    const project = await projectService.getProjectWithTasks(projectId, userId);
    
    if (!project) {
      return errorResponse(res, 'not found', 404);
    }

    return successResponse(res, project);
  } catch (err) {
    next(err);
  }
});

// PATCH /projects/:id
router.patch('/:id', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const projectId = req.params.id;
    const userId = (req as any).user.user_id;
    const validated = projectSchema.partial().parse(req.body);

    const result = await projectService.updateProject(projectId, userId, validated);
    if (result.error === 'not_found') return errorResponse(res, 'not found', 404);
    if (result.error === 'forbidden') return errorResponse(res, 'forbidden', 403);

    return successResponse(res, result.project);
  } catch (err) {
    next(err);
  }
});

// DELETE /projects/:id
router.delete('/:id', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const projectId = req.params.id;
    const userId = (req as any).user.user_id;

    const result = await projectService.deleteProject(projectId, userId);
    if (result.error === 'not_found') return errorResponse(res, 'not found', 404);
    if (result.error === 'forbidden') return errorResponse(res, 'forbidden', 403);

    return res.status(204).send();
  } catch (err) {
    next(err);
  }
});

export default router;

File: backend/internal/handlers/projects.go
package handlers

import (
	"encoding/json"
	"errors"
	"log/slog"
	"net/http"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/swaroop/taskflow/internal/database"
	"github.com/swaroop/taskflow/internal/middleware"
)

// ProjectHandler groups project-related HTTP handlers.
type ProjectHandler struct {
	db  *database.DB
	log *slog.Logger
}

// NewProjectHandler wires up the project handler.
func NewProjectHandler(db *database.DB, log *slog.Logger) *ProjectHandler {
	return &ProjectHandler{db: db, log: log}
}

// List handles GET /projects
//
// Optional query params: ?page=1&limit=20
func (h *ProjectHandler) List(w http.ResponseWriter, r *http.Request) {
	claims := middleware.ClaimsFrom(r)
	page, _ := strconv.Atoi(r.URL.Query().Get("page"))
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))

	result, err := h.db.ListProjects(r.Context(), database.ListProjectsParams{
		UserID: claims.UserID,
		Page:   page,
		Limit:  limit,
	})
	if err != nil {
		h.log.Error("list projects", "error", err)
		respondError(w, http.StatusInternalServerError, "internal server error")
		return
	}

	respondJSON(w, http.StatusOK, map[string]interface{}{
		"projects": result.Projects,
		"meta": map[string]interface{}{
			"page":     max(page, 1),
			"limit":    max(limit, 20),
			"total":    result.Total,
			"has_next": result.Total > max(page, 1)*max(limit, 20),
		},
	})
}

// Create handles POST /projects
//
// Request:  {"name":"...", "description":"..."}
// Response: 201 project object
func (h *ProjectHandler) Create(w http.ResponseWriter, r *http.Request) {
	claims := middleware.ClaimsFrom(r)

	var req struct {
		Name        string  `json:"name"`
		Description *string `json:"description"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		respondError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	req.Name = strings.TrimSpace(req.Name)
	if req.Name == "" {
		respondValidationError(w, map[string]string{"name": "is required"})
		return
	}

	project, err := h.db.CreateProject(r.Context(), req.Name, req.Description, claims.UserID)
	if err != nil {
		h.log.Error("create project", "error", err)
		respondError(w, http.StatusInternalServerError, "internal server error")
		return
	}

	respondJSON(w, http.StatusCreated, project)
}

// Get handles GET /projects/:id — returns project with embedded tasks.
func (h *ProjectHandler) Get(w http.ResponseWriter, r *http.Request) {
	id, err := parseUUID(chi.URLParam(r, "id"))
	if err != nil {
		respondError(w, http.StatusNotFound, "not found")
		return
	}

	project, err := h.db.GetProjectByID(r.Context(), id)
	if err != nil {
		if errors.Is(err, database.ErrNotFound) {
			respondError(w, http.StatusNotFound, "not found")
			return
		}
		h.log.Error("get project", "error", err)
		respondError(w, http.StatusInternalServerError, "internal server error")
		return
	}

	respondJSON(w, http.StatusOK, project)
}

// Update handles PATCH /projects/:id — owner only.
//
// Request:  {"name":"...", "description":"..."} (both optional)
func (h *ProjectHandler) Update(w http.ResponseWriter, r *http.Request) {
	claims := middleware.ClaimsFrom(r)

	id, err := parseUUID(chi.URLParam(r, "id"))
	if err != nil {
		respondError(w, http.StatusNotFound, "not found")
		return
	}

	// Verify ownership before any mutation
	existing, err := h.db.GetProjectByID(r.Context(), id)
	if err != nil {
		if errors.Is(err, database.ErrNotFound) {
			respondError(w, http.StatusNotFound, "not found")
			return
		}
		h.log.Error("get project for update", "error", err)
		respondError(w, http.StatusInternalServerError, "internal server error")
		return
	}
	if existing.OwnerID != claims.UserID {
		respondError(w, http.StatusForbidden, "forbidden")
		return
	}

	var req struct {
		Name        *string `json:"name"`
		Description *string `json:"description"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		respondError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	// Merge with existing values for fields not supplied
	newName := existing.Name
	if req.Name != nil {
		newName = strings.TrimSpace(*req.Name)
		if newName == "" {
			respondValidationError(w, map[string]string{"name": "cannot be empty"})
			return
		}
	}
	newDesc := existing.Description
	if req.Description != nil {
		newDesc = req.Description
	}

	updated, err := h.db.UpdateProject(r.Context(), id, newName, newDesc)
	if err != nil {
		h.log.Error("update project", "error", err)
		respondError(w, http.StatusInternalServerError, "internal server error")
		return
	}

	respondJSON(w, http.StatusOK, updated)
}

// Delete handles DELETE /projects/:id — owner only, cascades to tasks.
func (h *ProjectHandler) Delete(w http.ResponseWriter, r *http.Request) {
	claims := middleware.ClaimsFrom(r)

	id, err := parseUUID(chi.URLParam(r, "id"))
	if err != nil {
		respondError(w, http.StatusNotFound, "not found")
		return
	}

	existing, err := h.db.GetProjectByID(r.Context(), id)
	if err != nil {
		if errors.Is(err, database.ErrNotFound) {
			respondError(w, http.StatusNotFound, "not found")
			return
		}
		h.log.Error("get project for delete", "error", err)
		respondError(w, http.StatusInternalServerError, "internal server error")
		return
	}
	if existing.OwnerID != claims.UserID {
		respondError(w, http.StatusForbidden, "forbidden")
		return
	}

	if err := h.db.DeleteProject(r.Context(), id); err != nil {
		h.log.Error("delete project", "error", err)
		respondError(w, http.StatusInternalServerError, "internal server error")
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

// Stats handles GET /projects/:id/stats — bonus endpoint.
func (h *ProjectHandler) Stats(w http.ResponseWriter, r *http.Request) {
	id, err := parseUUID(chi.URLParam(r, "id"))
	if err != nil {
		respondError(w, http.StatusNotFound, "not found")
		return
	}

	// Ensure the project exists
	if _, err := h.db.GetProjectByID(r.Context(), id); err != nil {
		if errors.Is(err, database.ErrNotFound) {
			respondError(w, http.StatusNotFound, "not found")
			return
		}
		h.log.Error("get project for stats", "error", err)
		respondError(w, http.StatusInternalServerError, "internal server error")
		return
	}

	stats, err := h.db.GetProjectStats(r.Context(), id)
	if err != nil {
		h.log.Error("project stats", "error", err)
		respondError(w, http.StatusInternalServerError, "internal server error")
		return
	}

	respondJSON(w, http.StatusOK, stats)
}

// parseUUID is a small helper that returns a 404-safe error on bad input.
func parseUUID(s string) (uuid.UUID, error) {
	return uuid.Parse(s)
}

